From 9885c9ed7c4f4e6c7771b9c414d60ca9e881c541 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michal=20=C4=8Ciha=C5=99?= <michal@cihar.com>
Date: Sun, 22 Jul 2018 16:28:28 +0200
Subject: [PATCH 1/1] Add avatar cache cleanup management command
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This will make it easier to upgrade to Python 3.

Issue #2095

Signed-off-by: Michal Čihař <michal@cihar.com>
---
 docs/admin/management.rst                     | 10 ++++
 docs/admin/upgrade.rst                        |  2 +-
 .../commands/cleanup_avatar_cache.py          | 55 +++++++++++++++++++
 3 files changed, 66 insertions(+), 1 deletion(-)
 create mode 100644 weblate/accounts/management/commands/cleanup_avatar_cache.py

diff --git a/weblate/accounts/management/commands/cleanup_avatar_cache.py b/weblate/accounts/management/commands/cleanup_avatar_cache.py
new file mode 100644
index 000000000..f9bdd7a6e
--- /dev/null
+++ b/weblate/accounts/management/commands/cleanup_avatar_cache.py
@@ -0,0 +1,55 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright © 2012 - 2018 Michal Čihař <michal@cihar.com>
+#
+# This file is part of Weblate <https://weblate.org/>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+from glob import glob
+import pickle
+import os
+import zlib
+
+from django.conf import settings
+from django.core.cache import caches, InvalidCacheBackendError
+from django.core.management.base import BaseCommand
+
+
+class Command(BaseCommand):
+    help = 'removes incompatible files from avatar cache'
+
+    def handle(self, *args, **options):
+        backends = (
+            'django.core.cache.backends.filebased.FileBasedCache',
+        )
+        if 'avatar' not in settings.CACHES:
+            return
+        if settings.CACHES['avatar']['BACKEND'] not in backends:
+            return
+        mask = os.path.join(
+            settings.CACHES['avatar']['LOCATION'],
+            '*.djcache'
+        )
+        for name in glob(mask):
+            with open(name, 'rb') as handle:
+                try:
+                    # Load expiry
+                    pickle.load(handle)
+                    # Load payload
+                    pickle.loads(zlib.decompress(handle.read()))
+                except Exception as error:
+                    self.stdout.write('Removing {}: {}'.format(name, error))
+                    os.remove(name)
-- 
2.17.1

